### PR TITLE
Fix parsing of reserved keywords

### PR DESCRIPTION
**Pull request by patjakdev**
*Issue #, if available:* None

*Description of changes:*

This PR modifies the Cedar tokenizer to properly reserve the set of [reserved keywords](https://arc.net/l/quote/yvyhkfvf) from the Cedar spec. In addition, it resolves a parsing ambiguity for the `Primary` production, which requires a lookahead to determine whether it will be a `VAR`, `Entity`, or `ExtFun`.

### CODE REVIEW

**Comment by patjakdev** [comment]
See further discussion of this PR on the [StrongDM fork](https://github.com/strongdm/cedar-go/pull/7)

**Comment by kjamieson-sdm** [comment]
It looks like this forbids the use of reserved keywords in annotation keys, which the Rust implementation specifically allows (see https://github.com/cedar-policy/cedar/pull/634). Are we concerned about that difference in behaviour?

File: internal/parser/cedar_unmarshal_test.go [hunk 1]
```diff
@@ -584,3 +590,96 @@ func TestPolicySliceErrors(t *testing.T) {
...
+			nil, testutil.Error}, // Rust: unexpected token
+		{"scope/unless", `permit (principal == unless::"test", action, resource);`,
+			ast.Permit().PrincipalEq(types.NewEntityUID("unless", "test")), testutil.OK}, // Rust: OK
+		{"scope/when", `permit (principal == when::"test", action, resource);`,
+			ast.Permit().PrincipalEq(types.NewEntityUID("when", "test")), testutil.OK}, // Rust: OK
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+
```
Comment (kjamieson-sdm) [inline internal/parser/cedar_unmarshal_test.go:675]: Is `t.Parallel()` omitted intentionally?

**Comment by patjakdev** [comment]
> It looks like this forbids the use of reserved keywords in annotation keys, which the Rust implementation specifically allows (see [cedar-policy/cedar#634](https://github.com/cedar-policy/cedar/pull/634)). Are we concerned about that difference in behaviour?

Good catch @kjamieson-sdm. I was basing this on the [documented grammar](https://arc.net/l/quote/hzojfyne).

I'll add a test specifically for this case.

**Review by kjamieson-sdm — APPROVED** [review]

**Pull request merged**